### PR TITLE
Fix overflow problem (#874)

### PR DESCRIPTION
**Pull request by yairzaslavsky**
An overflow occured while calculating the expoential backoff.

as MAX_RETRIES is 30, with baseDelay of 5000 you will get to an overflow
at the calculation baseDelay * (1 << retriesAttempted), a negative value will
be returned from Math.min , and passed as the delay value to the call
to Thread.sleep, causing an exception.

My fix simply changes this to (1L << retriesAttempted) , which makes
the multiplication an a long value, avoiding the overflow

### CODE REVIEW

File: aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java [hunk 1]
```diff
@@ -103,7 +103,7 @@ public long delayBeforeNextRetry(AmazonWebServiceRequest originalRequest,
                                          AmazonClientException exception,
                                          int retriesAttempted) {
             return (retriesAttempted > MAX_RETRIES) ? maxBackoffTime :
-                    Math.min(((1 << retriesAttempted) * baseDelay), maxBackoffTime);
+                    Math.min(((1L << retriesAttempted) * baseDelay), maxBackoffTime);
```
Comment (kiiadi) [inline aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java]: Thanks for the PR - good catch. I think in terms of implementation though we're potentially just dodging the problem - as soon as you get MAX_RETRIES >= 63 we're going to have the same issue (ie: Long's overflow) - we might be better adding another call to Math to help us. This should resolve the issue:

`Math.max(1, 1 << retriesAttempted)`

Also we should use the same with all the pre-canned back-off strategies.

Comment (kiiadi) [inline aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java]: Alternatively we could do this which might actually be better:

```java
int exponential = 1 << retriesAttempted;
return exponential >= 0 ? exponential : Integer.MAX_VALUE;
```

Comment (yairzaslavsky) [inline aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java]: @kiiadi Thanks for your comments,
On 1.11.40 -  MAX_RERTIES is set to 30 , so I did not handle the potential overflow of long, as it does not reach the value of 63.

Using the exponential value  as an int and multiply it by baseDelay might still cause an Integer overflow.

Since the method dleayBeforeNextRetry returns long, I think it's better to resort to longs as soon as possible in the method's code.

I agree that other strategies should be fixed as well.

File: aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java [hunk 2]
```diff
@@ -77,7 +77,7 @@ public EqualJitterBackoffStrategy(final int baseDelay,
             this.maxBackoffTime = ValidationUtils.assertIsPositive(maxBackoffTime, "Max backoff");
         }
 
-        @Override
+        @Override*5
```
Comment (kiiadi) [inline aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java]: Is this a typo?

File: aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java [hunk 3]
```diff
@@ -102,8 +102,10 @@ public ExponentialBackoffStrategy(final int baseDelay,
         public long delayBeforeNextRetry(AmazonWebServiceRequest originalRequest,
                                          AmazonClientException exception,
                                          int retriesAttempted) {
+            long potentialWait = (1L << retriesAttempted) * baseDelay;
             return (retriesAttempted > MAX_RETRIES) ? maxBackoffTime :
-                    Math.min(((1 << retriesAttempted) * baseDelay), maxBackoffTime);
+                    (potentialWait < 0  ? maxBackoffTime :
```
Comment (kiiadi) [inline aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java]: Reckon we can push this into a static private method that the various strategies could use?

File: aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java (hunk 2 above)
Comment (yairzaslavsky) [inline aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java]: Yes.
Already suggested a fix.

File: aws-java-sdk-osgi/dependency-reduced-pom.xml [hunk 4]
```diff
@@ -3,7 +3,7 @@
   <parent>
     <artifactId>aws-java-sdk-pom</artifactId>
     <groupId>com.amazonaws</groupId>
-    <version>1.11.41</version>
+    <version>1.11.42-SNAPSHOT</version>
```
Comment (kiiadi) [inline aws-java-sdk-osgi/dependency-reduced-pom.xml]: Did you mean to commit changes to the pom file here? They seem unrelated.

File: aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java [hunk 5]
```diff
@@ -43,6 +43,13 @@
     /** Maximum retry limit.  Avoids integer overflow issues. **/
     private static final int MAX_RETRIES = 30;
 
+    private static final long exponentialDelayHelper(int retriesAttempted, int baseDelay, int maxBackoffTime) {
+      long potentialWait = (1L << retriesAttempted) * baseDelay;
+      return (retriesAttempted > MAX_RETRIES) ? maxBackoffTime :
+              (potentialWait < 0  ? maxBackoffTime :
+              Math.min(potentialWait, maxBackoffTime));
+    }
+
```
Comment (kiiadi) [inline aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java]: /Nitpick: do you mind if we push this method to the bottom of the class?

Also I suspect we can get away with just taking in "retriesAttempted" as a parameter since it's the `1 << retriesAttempted` line that we're trying to wrap. What if we did something like this:

```java
private static int calculateExponentialMultiplier(int retriesAttempted) {
    int multiplier = 1 << retriesAttempted;
    return multiplier >= 0 ? multiplier : Integer.MAX_VALUE;
}
```

File: aws-java-sdk-osgi/dependency-reduced-pom.xml (hunk 4 above)
Comment (yairzaslavsky) [inline aws-java-sdk-osgi/dependency-reduced-pom.xml]: My bad about the pom, thanks

File: aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java (hunk 5 above)
Comment (yairzaslavsky) [inline aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java]: I would prefer to stick to long here,  if you take the multiplier and multiply it by a baseDelay you will still have an overflow.
I would prefer to handle the overflow inside the helper method.

What do you think about this?

Comment (yairzaslavsky) [inline aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java]: Submitted a patch with no pom.xml change and moving the static method down in the file.

File: aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java [hunk 6]
```diff
@@ -147,4 +144,10 @@ public final long delayBeforeNextRetry(AmazonWebServiceRequest originalRequest,
         }
     }
 
-}
\ No newline at end of file
+    private static final long exponentialDelayHelper(int retriesAttempted, int baseDelay, int maxBackoffTime) {
```
Comment (kiiadi) [inline aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java]: /nitpick - can we rename this method so that it's a verb? Something like "calculateExponentialDelay"

Comment (yairzaslavsky) [inline aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java]: Looks like that as a part of running "mvn" locally the  pom.xml file gets changed.
I will rebase again against master and get rid of this change.

I am fine with the name suggestion.
